walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,78 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    Collection<TimeRange> meetingTimes = new LinkedHashSet<>();
+    int start = TimeRange.START_OF_DAY;
+    int end;
+
+    // Add event time to unavailable times if the event contains a required person.
+    for (Event event : events) {
+        if (!checkRequired(event.getAttendees(), request.getAttendees())) {
+            continue;
+        }
+        unavailableTimes.add(event.getWhen());
+    }
+
+    // Sort unavailable times by their start times.
+    Collections.sort(unavailableTimes);
+
+    // Loop through unavailable times to find potential meeting times
+    for (int i = 0; i < unavailableTimes.size(); i++) {
+      TimeRange time = unavailableTimes.get(i);
+
+      if (i < unavailableTimes.size() - 1) {
+        TimeRange nextTime = unavailableTimes.get(i + 1);
+
+        // Combine current and next times into one time.
+        if (time.overlaps(nextTime)) {
+            int earliest = Math.min(time.start(), nextTime.start());
+            int latest = Math.max(time.end(), nextTime.end());
+
+            time = TimeRange.fromStartEnd(earliest, latest, false);
+            i++;
+        }
+      }
+
+      // If the event starts in the beggining of the day, make the 
+      // first possible start at the end of this event. 
+      if (time.start() == 0) {
+        start = time.end();
+      }
+      end = time.start();
+
+      if (end - start > 0 && request.getDuration() <= end - start) {
+        meetingTimes.add(TimeRange.fromStartEnd(start, end, false));
+      }
+      start = time.end();
+    }
+
+    end = TimeRange.END_OF_DAY;
+
+    if (request.getDuration() <= end - start) {
+      meetingTimes.add(TimeRange.fromStartEnd(start, end, true));
+    }
```

Comment:
you shouldn't set `end` here. Just replace it with `TimeRange.END_OF_DAY`

```suggestion
    if (request.getDuration() <= TimeRange.END_OF_DAY - start) {
      meetingTimes.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
    }
```

Reply:
ah good idea